clang/lib/CodeGen/CGCall.cpp

```diff
@@ -2401,6 +2400,15 @@ void CodeGenModule::ConstructAttributeList(StringRef Name,
           FuncAttrs.addAttribute(llvm::Attribute::NoReturn);
         NBA = Fn->getAttr<NoBuiltinAttr>();
       }
+
+      auto FunctionEffects = Fn->getFunctionEffects();
+      for(size_t i = 0; i < FunctionEffects.size(); i++)
+      {
+        if(FunctionEffects.effects()[i].kind() == FunctionEffect::Kind::NonBlocking)
+        {
+          FuncAttrs.addAttribute(llvm::Attribute::NonBlocking);
+        }
+      }
```

Comment:
LOOK HERE!

In the past, all we had to do was look at if the attribute existed on the function. Now, doug uses these "function effects".

Please dig into the FunctionEffects class, and it's related classes and make sure you agree what I'm doing here is valid.


Try running a test program with a function marked [[clang::nonblocking]] [[clang::nonblocking(false)]] [[clang::nonblocking(true)]] 


Just in writing this comment, I discovered this does not catch [[clang::blocking(false)]] which it should. This may be something you could fix while you're in here

Reply:
Ok, [[clang::blocking(false)]] is now disallowed by the compiler, so I believe this code to be correct as written. The effect will "resolve" at this point to either blocking or nonblocking. `[[noblocking(false)]` resolves to `blocking` and so we never trip this conditional.